internal/weldr/api.go

```diff
@@ -1567,27 +1567,54 @@ func (api *API) blueprintsDepsolveHandler(writer http.ResponseWriter, request *h
 	common.PanicOnError(err)
 }
 
-// setPkgEVRA replaces the version globs in the blueprint with their EVRA values from the dependencies
-//
-// The dependencies must be pre-sorted for this function to work properly
-// It will return an error if it cannot find a package in the dependencies
-func setPkgEVRA(dependencies []rpmmd.PackageSpec, packages []blueprint.Package) error {
-	for pkgIndex, pkg := range packages {
-		i := sort.Search(len(dependencies), func(i int) bool {
-			return dependencies[i].Name >= pkg.Name
-		})
-		if i < len(dependencies) && dependencies[i].Name == pkg.Name {
-			if dependencies[i].Epoch == 0 {
-				packages[pkgIndex].Version = fmt.Sprintf("%s-%s.%s", dependencies[i].Version, dependencies[i].Release, dependencies[i].Arch)
-			} else {
-				packages[pkgIndex].Version = fmt.Sprintf("%d:%s-%s.%s", dependencies[i].Epoch, dependencies[i].Version, dependencies[i].Release, dependencies[i].Arch)
+// expandBlueprintGlobs will expand package name globs and versions using the depsolve results
+// The result is a sorted list of Package structs with the full package name and version
+// It will return an error if it cannot find a non-glob package name in the dependency list
+func expandBlueprintGlobs(dependencies []rpmmd.PackageSpec, packages []blueprint.Package) ([]blueprint.Package, error) {
+	newPackages := make(map[string]blueprint.Package)
+
+	for _, pkg := range packages {
+		if !strings.ContainsAny(pkg.Name, "*?") {
+			// No glob characters, find an exact match in dependencies
+			i := sort.Search(len(dependencies), func(i int) bool {
+				return dependencies[i].Name >= pkg.Name
+			})
+			if i >= len(dependencies) || dependencies[i].Name != pkg.Name {
+				// Packages should not be missing from the depsolve results
+				return nil, fmt.Errorf("%s missing from depsolve results", pkg.Name)
```

Comment:
No problem here, but this got me thinking if specifying virtual provides as a package name would work when building images, because that could potentially hit this condition (the virtual provides requested will depsolve to a different package name and the original will be missing from the list).

E.g. requesting `MTA` as a package may depsolve to `exim`, `postfix`, `sendmail`, etc... 

```toml
name = "test"
description = ""
version = "0.0.1"
modules = []
groups = []
distro = ""

[[packages]]
name = "MTA"
version = "*"
```
currently works

```toml
name = "test"
description = ""
version = "0.0.0"
modules = []
groups = []
distro = ""

[[packages]]
name = "MTA"
version = "*.*"
```
produces ```ERROR: BlueprintsError: test: DNF error occurred: MarkingErrors: Error occurred when marking packages for installation: Problems in request:
missing packages: MTA-*.*```. I'm not sure if specifying `*.*` version glob makes sense for virtual provides, but I wanted to point out this behavior.

Reply:
Yeah, I don't think that makes sense, since it ends up installing multiple packages. We should document it in the blueprint documentation though. There's really no way to figure out the mapping from installed packages back to the virtual provide.